### PR TITLE
Support SQL-compliant NaN ordering for DictionaryArray and non-DictionaryArray

### LINKED ISSUE #2599: Support SQL-compliant NaN ordering between for DictionaryArray and non-DictionaryArray
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Part of https://github.com/apache/arrow-rs/issues/2568.

The NaN ordering of comparison between DictionaryArray and non-DictionaryArray is not SQL-compliant. We should support SQL-compliant behavior.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2599.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
cc @sunchao

File: arrow/src/compute/kernels/comparison.rs [hunk 1]
```diff
@@ -2670,10 +2670,54 @@ pub fn eq_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                left,
+                right,
+                |a, b| a == b,
+                |a, b| a == b,
+                |a, b| {
+                    if is_nan(a) && is_nan(b) {
+                        true
+                    } else {
+                        a == b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:2693]: minor: 
```suggestion
(is_nan(a) && is_nan(b)) || (a == b)
```

File: arrow/src/compute/kernels/comparison.rs [hunk 2]
```diff
@@ -3084,10 +3312,54 @@ pub fn gt_eq_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                right,
+                left,
+                |a, b| a <= b,
+                |a, b| a <= b,
+                |a, b| {
+                    if is_nan(b) {
+                        true
+                    } else {
+                        a <= b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:3360]: minor: 
```suggestion
is_nan(b) || (a <= b)
```

File: arrow/src/compute/kernels/comparison.rs [hunk 3]
```diff
@@ -3084,10 +3312,54 @@ pub fn gt_eq_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                left,
+                right,
+                |a, b| a >= b,
+                |a, b| a >= b,
+                |a, b| {
+                    if is_nan(a) {
+                        true
+                    } else {
+                        a >= b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:3335]: minor:
```suggestion
is_nan(a) || (a >= b)
```

File: arrow/src/compute/kernels/comparison.rs [hunk 4]
```diff
@@ -2670,10 +2670,54 @@ pub fn eq_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                right,
+                left,
+                |a, b| a == b,
+                |a, b| a == b,
+                |a, b| {
+                    if is_nan(a) && is_nan(b) {
+                        true
+                    } else {
+                        a == b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:2718]: minor
```suggestion
(is_nan(a) && is_nan(b)) || a == b)
```

File: arrow/src/compute/kernels/comparison.rs [hunk 5]
```diff
@@ -2752,10 +2796,54 @@ pub fn neq_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                right,
+                left,
+                |a, b| a != b,
+                |a, b| a != b,
+                |a, b| {
+                    if is_nan(a) && is_nan(b) {
+                        false
+                    } else {
+                        a != b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:2844]: minor
```suggestion
!(is_nan(a) && is_nan(b)) && (a != b)
```

File: arrow/src/compute/kernels/comparison.rs [hunk 6]
```diff
@@ -2752,10 +2796,54 @@ pub fn neq_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                left,
+                right,
+                |a, b| a != b,
+                |a, b| a != b,
+                |a, b| {
+                    if is_nan(a) && is_nan(b) {
+                        false
+                    } else {
+                        a != b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:2819]: minor
```suggestion
!(is_nan(a) && is_nan(b)) && (a != b)
```

File: arrow/src/compute/kernels/comparison.rs [hunk 7]
```diff
@@ -2836,10 +2924,58 @@ pub fn lt_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                |a, b| a < b,
+                |a, b| a < b,
+                |a, b| {
+                    if is_nan(a) {
+                        false
+                    } else if is_nan(b) {
+                        true
+                    } else {
+                        a < b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:2949]: minor
```suggestion
!is_nan(a) && (is_nan(b) || (a < b))
```

File: arrow/src/compute/kernels/comparison.rs [hunk 8]
```diff
@@ -2919,10 +3055,54 @@ pub fn lt_eq_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                right,
+                left,
+                |a, b| a >= b,
+                |a, b| a >= b,
+                |a, b| {
+                    if is_nan(b) {
+                        is_nan(a)
+                    } else {
+                        a >= b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:3103]: minor
```suggestion
(is_nan(b) && is_nan(a)) || (!is_nan(b) && (a >= b))
```

File: arrow/src/compute/kernels/comparison.rs [hunk 9]
```diff
@@ -3002,10 +3182,58 @@ pub fn gt_dyn(left: &dyn Array, right: &dyn Array) -> Result<BooleanArray> {
...
+                |a, b| a < b,
+                |a, b| a < b,
+                |a, b| {
+                    if is_nan(b) {
+                        !is_nan(a)
+                    } else if is_nan(a) {
+                        false
+                    } else {
+                        a < b
+                    }
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/comparison.rs:3234]: minor
```suggestion
(is_nan(b) && !is_nan(a) ) ||
(!is_nan(b) && !is_nan(a) && (a < b))
```

Comment (viirya) [reply arrow/src/compute/kernels/comparison.rs:3234]: Actually I'm not a fan of these shorten expressions. It is more hard to read and easier to mistake.

**Comment by viirya** [comment]
@tustvold As I'm going to use total_cmp, how about we merge this first in, and then I propose a PR to change to total_cmp and clean up the feature flag?

File: arrow/src/compute/kernels/comparison.rs (hunk 9 above)
Comment (sunchao) [reply arrow/src/compute/kernels/comparison.rs:3234]: Agree. It's more readable this way.

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
I'm going to merge this in and make a PR to use total_cmp and remove the flag.

**Pull request merged**